Mathlib/FieldTheory/NormalizedTrace.lean

```diff
@@ -0,0 +1,148 @@
+      normalizedTraceAux_intermediateField F K b',
+      normalizedTraceAux_intermediateField F K ab',
+      normalizedTraceAux_eq_of_fininteDimensional F a',
+      normalizedTraceAux_eq_of_fininteDimensional F b',
+      normalizedTraceAux_eq_of_fininteDimensional F ab',
+      ← add_div, ← map_add]
+    rfl
+  map_smul' m a := by
+    dsimp only [AddHom.toFun_eq_coe, AddHom.coe_mk, RingHom.id_apply]
+    let E := F⟮a⟯
+    haveI : FiniteDimensional F F⟮a⟯ := adjoin.finiteDimensional (Algebra.IsIntegral.isIntegral a)
+    have ha : a ∈ E := mem_adjoin_simple_self F a
+    have hma : m • a ∈ E := smul_mem E ha
+    let a' : E := ⟨a, ha⟩
+    let ma' : E := ⟨m • a, hma⟩
+    rw [normalizedTraceAux_intermediateField F K a',
+      normalizedTraceAux_intermediateField F K ma',
+      normalizedTraceAux_eq_of_fininteDimensional F a',
+      normalizedTraceAux_eq_of_fininteDimensional F ma',
+      ← smul_div_assoc, ← map_smul]
+    rfl
+
+variable {F} in
+theorem normalizedTrace_self_apply (a : F) : normalizedTrace F F a = a := by
+  dsimp [normalizedTrace]
+  rw [normalizedTraceAux_eq_of_fininteDimensional F a, Module.finrank_self F,
+    Nat.cast_one, div_one, Algebra.trace_self_apply]
+
+@[simp]
+theorem normalizedTrace_self : normalizedTrace F F = LinearMap.id :=
+  LinearMap.ext normalizedTrace_self_apply
+
+@[simp]
+theorem normalizedTrace_algebraMap_apply (a : F) : normalizedTrace F K (algebraMap F K a) = a :=
+  (Algebra.ofId_apply K a ▸
+    normalizedTraceAux_map F K (Algebra.ofId F K) a).trans (normalizedTrace_self_apply a)
+
+/-- The normalized trace map is a left inverse of the algebra map. -/
+@[simp]
+theorem normalizedTrace_algebraMap : normalizedTrace F K ∘ algebraMap F K = id :=
+  funext <| normalizedTrace_algebraMap_apply F K
+
+/-- The normalized trace transfers via (injective) maps. -/
+@[simp]
+theorem normalizedTrace_map {E : Type*} [Field E] [Algebra F E] [Algebra.IsIntegral F E]
+    (f : E →ₐ[F] K) (a : E) : normalizedTrace F K (f a) = normalizedTrace F E a :=
+  normalizedTraceAux_map F K f a
+
+/-- The normalized trace commutes with (injective) maps. -/
+@[simp]
+theorem normalizedTrace_comp {E : Type*} [Field E] [Algebra F E] [Algebra.IsIntegral F E]
+    (f : E →ₐ[F] K) : normalizedTrace F K ∘ₗ f = normalizedTrace F E :=
+  LinearMap.ext <| normalizedTrace_map F K f
+
+/-- The normalized trace transfers via restriction to a subextension. -/
+theorem normalizedTrace_intermediateField {E : IntermediateField F K} (a : E) :
+    normalizedTrace F K a = normalizedTrace F E a :=
+  normalizedTraceAux_intermediateField F K a
+
+theorem normalizedTrace_eq_of_fininteDimensional [FiniteDimensional F K] :
+    normalizedTrace F K = (Module.finrank F K : F)⁻¹ • Algebra.trace F K := by
+  ext x
+  rw [LinearMap.smul_apply, smul_eq_mul, mul_comm, ← div_eq_mul_inv]
+  apply normalizedTraceAux_eq_of_fininteDimensional
+
+/-- The normalized trace map is non-trivial. -/
+theorem nontrivial_normalizedTrace : normalizedTrace F K ≠ 0 :=
```

Comment:
Maybe this should called `normalizedTrace_ne_zero` as `nontrivial` in theorem name usually refers to Lean's `Nontrivial`.